File: 416-partition-equal-subset-sum/416-partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = Arrays.stream(nums).sum();
        if(sum%2 != 0) return false;
        int halfSum = sum / 2;
        
        Boolean[][] dp = new Boolean[nums.length][halfSum + 1];
        
        return helper(nums,0,halfSum,dp);
        
        
    }
    
    private boolean helper(int[] nums, int idx,int sum, Boolean[][] dp){
        if(sum == 0) return true; 
        else if(idx >= nums.length || sum < 0) return false;
        else if(dp[idx][sum] != null) return dp[idx][sum];
        
        return dp[idx][sum] =   helper(nums,idx + 1, sum - nums[idx],dp) ||
        helper(nums, idx + 1, sum, dp);
        
    }
}

File: 416-partition-equal-subset-sum/416-partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        n = len(nums)
        total = sum(nums)
        if(total % 2 != 0):
            return False
        half = total // 2
        
        @lru_cache(None)
        def helper(i,curr):
            if curr < 0 or i >= n:
                return False
            elif curr == 0:
                return True
            
            return helper(i + 1, curr - nums[i]) or helper(i+1,curr)
        
        return helper(0,half)